hasktorch/src/Torch/Typed/NN/Recurrent/LSTM.hs

```diff
@@ -0,0 +1,478 @@
+    -> [D.Tensor]
+lstmParamsToTlist' acc (LSTMLayerK l wi wh bi bh) =
+    lstmParamsToTlist' acc l
+        <> ( (toDynamic $ toDependent wi)
+           : (toDynamic $ toDependent wh)
+           : (toDynamic $ toDependent bi)
+           : (toDynamic $ toDependent bh)
+           : []
+           )
+lstmParamsToTlist' acc (LSTMLayer1 wi wh bi bh) =
+    (toDynamic $ toDependent wi)
+        : (toDynamic $ toDependent wh)
+        : (toDynamic $ toDependent bi)
+        : (toDynamic $ toDependent bh)
+        : acc
+lstmParamsToTlist l = lstmParamsToTlist' [] l
+
+ziplstmLayers :: [D.Tensor] -> [D.Tensor] -> [D.Tensor] -> [D.Tensor]
+ziplstmLayers acc [] [] = acc
+ziplstmLayers acc (a : b : c : d : xs) (a' : b' : c' : d' : xs') =
+    a : b : c : d : a' : b' : c' : d' : ziplstmLayers acc xs xs'
+
+forward'
+    :: forall device dtype dir numLayers inputDim hiddenDim seqLen batchSize
+     . ( KnownNat numLayers
+       , KnownNat inputDim
+       , KnownNat hiddenDim
+       , KnownNat seqLen
+       , KnownNat batchSize
+       , KnownNat (NumberOfDirections dir)
+       )
+    => Bool
+    -> LSTM device dir dtype numLayers inputDim hiddenDim
+    -> Tensor device dtype '[seqLen, batchSize, inputDim]
+    -> ( Tensor
+             device
+             dtype
+             '[seqLen, batchSize, hiddenDim * NumberOfDirections dir]
+       , Tensor
+             device
+             dtype
+             '[numLayers * NumberOfDirections dir, batchSize, hiddenDim]
+       , Tensor
+             device
+             dtype
+             '[numLayers * NumberOfDirections dir, batchSize, hiddenDim]
+       )
+forward' doDropout (LSTM p (Dropout d) cc hc) input =
+    lstm @numLayers @(NumberOfDirections dir) input
+                                              (cc', hc')
+                                              (params p)
+                                              d
+                                              doDropout
+  where
+    cc' =
+        reshape @'[numLayers * (NumberOfDirections dir), batchSize, hiddenDim]
+            $ expand
+                  @'[batchSize, numLayers * (NumberOfDirections dir), hiddenDim]
+                  False -- TODO: What does the bool do?
+                  cc
+    hc' =
+        reshape @'[numLayers * (NumberOfDirections dir), batchSize, hiddenDim]
+            $ expand
+                  @'[batchSize, numLayers * (NumberOfDirections dir), hiddenDim]
+                  False
+                  hc
+params
+    :: forall device dtype inputDim hiddenDim numLayers dir
+     . ParamsPerDirection device dtype inputDim hiddenDim numLayers dir
+    -> [D.Tensor]
+params (BidirectionalParams fwd rvs) =
+    ziplstmLayers [] (lstmParamsToTlist' [] fwd) (lstmParamsToTlist' [] rvs)
+params (UniDirectionalParams fwd) = lstmParamsToTlist' [] fwd
+
+-- | Forward propagage the `LSTM` module and apply dropout on the outputs of each layer.
+--
+-- >>> input ::Tensor 'D.Float '[5,16,10]          <- randn
```

Comment:
that should be `CPUTensor 'D.Float '[5,16,10]`.
are you sure, doctests runs on this file?

Reply:
I thought the doctests were auto-discovered but since found the file where they're specified, let's see if it works now.